### DISCUSSION TITLE
[help] Proposed way to integrate retry & timeout into {targets} with {crew} + output printing

### DISCUSSION THREAD

**Discussion by mzorko** [ropensci/targets#1066] [Help] [answered]
### Help

- [x] I understand and agree to <https://books.ropensci.org/targets/help.html>.

### Description

Hi all,

We are implementing {`targets`} with {`crew`} library (currently locally, but with plans to think about AWS ECS/EKS computing).

Is there a proposed way to set-up timeout and number of retries for each target separately?

We could wrap-up the code using `R.utils::withTimeout()` & some simple while loop for retries, but we are not sure if that is the right way to do it.

Also, is there a way to print out logger/print output from `targets` when making the pipeline using {`crew`}?

Thanks,
Mislav

**Comment by wlandau** [comment]
What kind of scenario are you envisioning for retries and timeouts? What kinds of tasks to you need to retry?

You could use `R.utils::withTimeout()`, ~~or alternatively [`purrr::insistently()`](https://purrr.tidyverse.org/reference/insistently.html) is another way.~~

For logging, that will depend on the platform. For local files, you could use something like [`futile.logger`](https://github.com/zatonovo/futile.logger) inside your targets and functions. For workflows with [`crew.cluster`](https://wlandau.github.io/crew.cluster/), each plugin has an argument like `slurm_log_output` (in the case of the [SLURM controller](https://wlandau.github.io/crew.cluster/reference/crew_controller_slurm.html)) to control where the output log files go. I haven't started implementing plugins for AWS compute just yet, but I think logging is probably going to involve CloudWatch in some way.

**Reply by wlandau** [reply]
Actually, `purrr::insistently()` may not be a good fit for `targets` because of how it interferes with dependency discovery, c.f. https://github.com/ropensci/targets/issues/638.

**Reply by mzorko** [reply]
Hi @wlandau, 

Thanks for your guidance. We will try to implement using `R.utils::withTimeout()` although as far as we know it is not robust when it comes to code ran outside of R (e.g., using cpp in some cases). 

On the logging part, we are using `logger` library to produce them, but when we run `tar_make()` using `crew::crew_controller_local()` they are not printed. Also, all `print(...)` lines are not printed as well in the console - we only see target started/skip/done lines. If we run it without `crew`, it works perfectly.

**Reply by wlandau** [reply, marked as answer]
I did some more digging, and I actually recommend looking into the {retry} package: https://github.com/randy3k/retry.

{targets} captures warnings and errors, but anything else written to stdout or stderr will only print to the R console if it does not run in a parallel worker. To see those print messages,  they will need to be written to a physical file. (When {crew} runs on cloud services, logging may work a bit differently, but {targets} will still not be able to pick up the messages of external processes and print them on the main controlling R process. )

**Reply by mzorko** [reply]
Thanks @wlandau! We will try it out.
